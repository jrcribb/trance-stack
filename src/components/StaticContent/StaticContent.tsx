import { createElement, useRef, useState, useEffect } from 'react';
import type { ReactNode } from 'react';

const useStaticContent = () => {
  const ref = useRef<HTMLElement>(null);
  const [render, setRender] = useState(typeof document === 'undefined');

  useEffect(() => {
    // check if the innerHTML is empty as client side navigation
    // need to render the component without server-side backup
    const isEmpty = ref.current?.innerHTML === '';
    if (isEmpty) {
      setRender(true);
    }
  }, []);

  return [render, ref];
};

export const StaticContent = <Elem extends keyof JSX.IntrinsicElements = 'div'>(
  { children, element, ...props }: {
    element?: Elem;
    children?: ReactNode;
  } & JSX.IntrinsicElements[Elem]
) => {
  const elem = element || 'div';
  const [render, ref] = useStaticContent();
  console.log('render');
  // if we're in the server or a spa navigation, just render it
  if (render) {
    console.log('server');
    return createElement(elem, {
      ...props,
      children: children
    });
  }
  console.log('client');
  // avoid re-render on the client
  return createElement(elem, {
    ...props,
    ref: ref,
    suppressHydrationWarning: true,
    dangerouslySetInnerHTML: { __html: '' }
  });
};
