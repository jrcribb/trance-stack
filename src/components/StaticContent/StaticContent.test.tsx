import react from 'react';
import { render } from '@testing-library/react';
import { expect, it, vi } from 'vitest';
import { StaticContent } from './StaticContent';

describe('StaticContent', () => {
  it('should wrap with div by default', () => {
    const { container } = render(<StaticContent title={'some div'}>{'div content'}</StaticContent>);
    expect(container).toMatchSnapshot();
  });

  it('should wrap with given element', () => {
    const component = <StaticContent async element={'script'} nonce={'example-nonce'}
      dangerouslySetInnerHTML={{
        __html: 'alert("hello world")'
      }}/>;

    const { container } = render(component);
    expect(container).toMatchSnapshot();

  });
});

