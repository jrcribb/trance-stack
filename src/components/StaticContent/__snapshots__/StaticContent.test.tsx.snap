// Vitest Snapshot v1

exports[`StaticContent > should wrap with div by default 1`] = `
<div>
  <div
    title="some div"
  >
    div content
  </div>
</div>
`;

exports[`StaticContent > should wrap with given element 1`] = `
<div>
  <script
    async=""
    nonce="example-nonce"
  >
    alert("hello world")
  </script>
</div>
`;

exports[`StaticContent > should wrap with given element 2`] = `
<div>
  <script
    async=""
    nonce="example-nonce"
  >
    alert("hello world")
  </script>
</div>
`;
